cpu/stm32/periph/adc_f3.c

```diff
@@ -0,0 +1,164 @@
+#if defined(RCC_AHBENR_ADC1EN)
+    periph_clk_en(AHB, (RCC_AHBENR_ADC1EN << adc_config[line].dev));
+#endif
+}
+
+static inline void done(adc_t line)
+{
+/* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
+ * don't disable the clock as the other device may still use it. */
+#if defined(RCC_AHBENR_ADC1EN)
+    periph_clk_dis(AHB, (RCC_AHBENR_ADC1EN << adc_config[line].dev));
+#endif
+    mutex_unlock(&locks[adc_config[line].dev]);
+}
+
+int adc_init(adc_t line)
+{
+    /* Check if the line is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+
+    /* Lock device and enable its peripheral clock */
+    prep(line);
+/* On some STM32F3 ADC are grouped by paire (ADC12EN or ADC34EN) so
+ * enable the clock only once here. */
+#if !defined(RCC_AHBENR_ADC1EN)
+    periph_clk_en(AHB, (RCC_AHBENR_ADC12EN << adc_config[line].dev));
+#endif
+
+    /* Setting ADC clock to HCLK/1 is only allowed if AHB clock
+     * prescaler is 1 */
+    if (!(RCC->CFGR & RCC_CFGR_HPRE_3)) {
+        /* set ADC clock to HCLK/1 */
+        ADC12_COMMON->CCR |= ADC_CCR_CKMODE_0;
+    }
+    else {
+        /* set ADC clock to HCLK/2 otherwise */
+        ADC12_COMMON->CCR |= ADC_CCR_CKMODE_1;
+    }
+
+    /* Configure the pin */
+    gpio_init_analog(adc_config[line].pin);
+
+    /* Init ADC line only if it wasn't already initialized */
+    if (!(dev(line)->CR & ADC_CR_ADEN)) {
+        /* Enable ADC internal voltage regulator and wait for startup period */
+        dev(line)->CR |= ADC_CR_ADVREGEN;
+        xtimer_usleep(ADC_T_ADCVREG_STUP_US);
+
+        /* Configure calibration for single ended input */
+        dev(line)->CR &= ~ADC_CR_ADCALDIF;
+
+        /* Start automatic calibration and wait for it to complete */
+        dev(line)->CR |= ADC_CR_ADCAL;
+        while (dev(line)->CR  & ADC_CR_ADCAL) {}
+
+        /* Clear ADRDY by writing it */
+        dev(line)->ISR |= ADC_ISR_ADRDY;
+
+        /* Enable ADC and wait for it to be ready */
+        dev(line)->CR |= ADC_CR_ADEN;
+        while ((dev(line)->ISR & ADC_ISR_ADRDY) == 0) {}
+
+        /* Set sequence length to 1 conversion */
+        dev(line)->SQR1 |= (0 & ADC_SQR1_L);
+    }
+
+    /* Configure sampling time for the given channel,
+     * set to 4.5 ADC clock cycles for slow channels */
+#if defined(CPU_MODEL_STM32F334R8) || defined(CPU_MODEL_STM32F302R8)
+    dev(line)->SMPR1 = (0x2 << ADC_SMPR1_SMP6_Pos)
+                        | (0x2 << ADC_SMPR1_SMP7_Pos);
+    dev(line)->SMPR2 = (0x2 << ADC_SMPR2_SMP11_Pos);
+#elif defined(CPU_MODEL_STM32F303RE)
+    dev(line)->SMPR1 = (0x2 << ADC_SMPR1_SMP6_Pos)
+                        | (0x2 << ADC_SMPR1_SMP7_Pos);
+    dev(line)->SMPR2 = (0x2 << ADC_SMPR2_SMP12_Pos);
+#elif defined(CPU_MODEL_STM32F303ZE)
+    dev(line)->SMPR1 = (0x2 << ADC_SMPR1_SMP6_Pos)
+                        | (0x2 << ADC_SMPR1_SMP8_Pos);
+                        | (0x2 << ADC_SMPR1_SMP9_Pos);
+    dev(line)->SMPR2 = (0x2 << ADC_SMPR2_SMP10_Pos);
+#endif
```

Comment:
Why is this different on different members of the family?

Reply:
Because of nucleo 32 / 64 / 144 arduino pins routing... you are right, that was a bad idea.

Better to do it for all slow channels (>5).